gv.c

```diff
@@ -721,55 +721,92 @@ S_maybe_add_coresub(pTHX_ HV * const stash, GV *gv,
 
 /*
 =for apidoc      gv_fetchmeth
+=for apidoc_item gv_fetchmeth_autoload
 =for apidoc_item gv_fetchmeth_pv
+=for apidoc_item gv_fetchmeth_pv_autoload
 =for apidoc_item gv_fetchmeth_pvn
+=for apidoc_item gv_fetchmeth_pvn_autoload
 =for apidoc_item gv_fetchmeth_sv
+=for apidoc_item gv_fetchmeth_sv_autoload
 
 These each look for a glob with name C<name>, containing a defined subroutine,
 returning the GV of that glob if found, or C<NULL> if not.
 
-C<stash> is always searched (first), unless it is C<NULL>.
+Searching is always done in the following order, with some steps skipped
+depending on various criteria.  The first match found is used, ending the
+search.  C<gv_fetchmeth_pv> and C<gv_fetchmeth_pv_autoload> lack a flags
+parameter, so it effectively is zero in the following.
 
-If C<stash> is NULL, or was searched but nothing was found in it, and the
-C<GV_SUPER> bit is set in C<flags>, stashes accessible via C<@ISA> are searched
-next.  Searching is conducted according to L<C<MRO> order|perlmroapi>.
+=over
+
+=item 1
+
+C<stash> is always searched first, unless it is C<NULL>.
+
+=item 2
+
+If the C<GV_SUPER> bit is set in C<flags>, stashes accessible via C<@ISA> are
+searched next.  Searching is conducted according to L<C<MRO> order|perlmroapi>.
+
```

Comment:
I don't think this is correct.

If GV_SUPER isn't set, stash is searched

If not found, @ISA is searched in MRO.

Reply:
My modus operandi for doing these pod changes is to assume the previous documentation is correct.  If something doesn't add up, I go look lat the source; much of which I'm unfamiliar with.

In this case, the following lines seem to be the crux
 ```
  if (flags & GV_SUPER) {
        if (!HvAUX(stash)->xhv_mro_meta->super)
            HvAUX(stash)->xhv_mro_meta->super = newHV();
        cachestash = HvAUX(stash)->xhv_mro_meta->super;
    }
    else cachestash = stash;
    /* check locally for a real method or a cache entry */
    he = (HE*)hv_common(
        cachestash, meth, name, len, is_utf8 ? HVhek_UTF8 : 0, create, NULL, 0

```

That looks to me without delving deeper that when the flag is set it does a `newHV` if necessary and sets `cachestash` to it, and then goes looking there.  So what am I getting wrong?

Reply:
> My modus operandi for doing these pod changes is to assume the previous documentation is correct. If something doesn't add up, I go look lat the source; much of which I'm unfamiliar with.

a1f822c22a9 made the following change to this documentation:
```
-C<GV_SUPER> indicates that we want to look up the method in the superclasses
-of the C<stash>.
+If C<stash> is NULL, or was searched but nothing was found in it, and the
+C<GV_SUPER> bit is set in C<flags>, stashes accessible via C<@ISA> are searched
+next.  Searching is conducted according to L<C<MRO> order|perlmroapi>.
```

The original there was ambiguous, but unfortunately the edit got the meaning wrong.  GV_SUPER means only the classes in `@ISA` are searched, skipping the supplied stash, rather then the default of searching stash and then the `@ISA` classes.

> So what am I getting wrong?

As you show gv_fetchmeth_internal() starts by setting cachestash to either the specified stash (no `GV_SUPER`), or the super cache HV for the stash, creating it if necessary (with `GV_SUPER`).

So if `GV_SUPER` is clear, stash is searched.

If `GV_SUPER` is set, the super cache is searched, *not* the specified stash.

If that lookup succeeds:
```
    if (he) gvp = (GV**)&HeVAL(he);
    else gvp = NULL;

    if(gvp) {
 ... check for a valid non-cache CV, or for a non-stale cache entry and succeed
 ... note this is in stash for no GV_SUPER, in the super cache with GV_SUPER
    }
```
If that fails we search the linearized ISA tree of stash (not cachestash), whether or not GV_SUPER is set:
```
    linear_av = mro_get_linear_isa(stash); /* has ourselves at the top of the list */
    linear_svp = AvARRAY(linear_av) + 1; /* skip over self */
    items = AvFILLp(linear_av); /* no +1, to skip over self */
    while (items--) {
        linear_sv = *linear_svp++;
        assert(linear_sv);
        cstash = gv_stashsv(linear_sv, 0);
...
    }
```